src/fs/bundle.rs

```diff
@@ -0,0 +1,235 @@
+use crate::fs::{FsNode, GuestPath};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use std::fmt::Debug;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+use zip::result::ZipError;
+use zip::ZipArchive;
+
+/// A helper struct to build an FsNode with files and directories coming in arbitrary order.
+/// This is required, because ZIP files are allowed to store entries in arbitrary order.
+struct FsNodeBuilder {
+    root: FsNode,
+}
+
+impl FsNodeBuilder {
+    fn new() -> Self {
+        Self {
+            root: FsNode::dir(),
+        }
+    }
+
+    fn find_or_make_directory(&mut self, path: &GuestPath) -> &mut FsNode {
```

Comment:
Just curious, is this because the entry for `/SomeDir/SomeFoo` can appear before the entry for `/SomeDir` or something like that?

Reply:
Yeah, I think this is it. The zip spec says "files can come in any order", extraction logic in `zip` crate always tries to create parent directories, so I guess this can be possible (even though it's weird)